server.go

```diff
@@ -234,3 +235,94 @@ func (b *backend) Move(r *http.Request, dest *internal.Href, overwrite bool) (cr
 	}
 	return created, err
 }
+
+// BackendSuppliedHomeSet represents either a CalDAV calendar-home-set or a
+// CardDAV addressbook-home-set. It should only be created via
+// `caldav.NewCalendarHomeSet()` or `carddav.NewAddressbookHomeSet()`. Only to
+// be used server-side, for listing a user's home sets as determined by the
+// (external) backend.
+type BackendSuppliedHomeSet interface {
+	GetXMLName() xml.Name
+}
+
+// UserPrincipalBackend can determine the current user's principal URL for a
+// given request context.
+type UserPrincipalBackend interface {
+	CurrentUserPrincipal(ctx context.Context) (string, error)
+}
+
+type ServeUserPrincipalOptions struct {
+	UserPrincipalPath string
+	HomeSets          []BackendSuppliedHomeSet
+}
+
+// ServeUserPrincipal replies to discovery requests. It returns false if the
+// request wasn't handled.
+func ServeUserPrincipal(w http.ResponseWriter, r *http.Request, options ServeUserPrincipalOptions) bool {
+	if r.Method != "PROPFIND" {
```

Comment:
Hmm. If `r.URL.Path != b.options.UserPrincipalPath` we shouldn't handle the request, right? Should we return false here in that case?

Reply:
(In which case, maybe we can just let the caller perform the path check and drop `UserPrincipalPath`?)

Reply:
I thought the idea was that this handles everything up to, including the user principal URL. I've observed at least Evolution sending PROPSTAT to `/` during discovery, which then returns the user principal URL.

Reply:
That's also why the .well-know handling was in there.

Reply:
Hm I'd rather just handle the principal stuff in here (as the function name suggests).

Reply:
Ok, I rewrote that function based on #74 (feel free to add yourself as co-author :slightly_smiling_face: ). I didn't take everything from there to not blow this up any more, I hope that's fine. Happy to add more bits and pieces afterwards.